src/common_subset.rs

```diff
@@ -105,25 +133,47 @@ impl<NodeUid: Clone + Debug + Display + Eq + Hash + Ord> CommonSubset<NodeUid> {
         }
     }
 
+    /// Receives a message form a remote node `sender_id`, and returns an optional result of the
+    /// Common Subset algorithm - a set of proposed values - and a queue of messages to be sent to
+    /// remote nodes, or an error.
+    pub fn handle_message(
+        &mut self,
+        sender_id: &NodeUid,
+        message: Message<NodeUid>,
+    ) -> Result<CommonSubsetOutput<NodeUid>, Error> {
+        match message {
+            Message::Broadcast(p_id, b_msg) => self.handle_broadcast(sender_id, &p_id, b_msg),
+            Message::Agreement(p_id, a_msg) => self.handle_agreement(sender_id, &p_id, &a_msg),
+        }
+    }
+
     /// Receives a broadcast message from a remote node `sender_id` concerning a
     /// value proposed by the node `proposer_id`. The output contains an
     /// optional result of the Common Subset algorithm - a set of proposed
     /// values - and a queue of messages to be sent to remote nodes, or an
     /// error.
-    pub fn handle_broadcast(
+    fn handle_broadcast(
         &mut self,
         sender_id: &NodeUid,
         proposer_id: &NodeUid,
         bmessage: BroadcastMessage,
     ) -> Result<CommonSubsetOutput<NodeUid>, Error> {
         let mut instance_result = None;
-        let input_result: Result<VecDeque<Output<NodeUid>>, Error> = {
-            if let Some(broadcast_instance) = self.broadcast_instances.get(proposer_id) {
+        let input_result: Result<
+            VecDeque<TargetedMessage<Message<NodeUid>, NodeUid>>,
+            Error,
+        > = {
+            if let Some(broadcast_instance) = self.broadcast_instances.get_mut(proposer_id) {
                 broadcast_instance
                     .handle_broadcast_message(sender_id, bmessage)
                     .map(|(opt_value, queue)| {
                         instance_result = opt_value;
-                        queue.into_iter().map(Output::Broadcast).collect()
+                        queue
+                            .into_iter()
+                            .map(|msg| {
+                                msg.map(|b_msg| Message::Broadcast(proposer_id.clone(), b_msg))
```

Comment:
Looks a bit involved but I guess this is how adding the proposer to the message was supposed to work anyway.

Reply:
I'm hoping that the `TargetedMessage` and its `map` method makes the distinction explicit between the message content, which higher-level algorithms will probably wrap into their own message types, and the addressee, to whom the wrapped message needs to be sent.
But maybe we can find a better way to do that.